Issue 233: Make intermediate-level grouping functions

kaitejohnson opened the pull request:
## Description

This PR closes #233. It creates two new intermediate level grouping functions:
1. Estimates and applies a delay from a reporting triangle 
2. Estimates and applies uncertainty from a reporting triangle and a point nowcast matrix. 


## Checklist

- [ ] My PR is based on a package issue and I have explicitly linked it.
- [ ] I have included the target issue or issues in the PR title in the for Issue(s) *issue-numbers*: PR title
- [ ] I have read the [contribution guidelines](https://github.com/epinowcast/.github/blob/main/CONTRIBUTING.md).
- [ ] I have tested my changes locally.
- [ ] I have added or updated unit tests where necessary.
- [ ] I have updated the documentation if required.
- [ ] My code follows the established coding standards.
- [ ] I have added a news item linked to this PR.
- [ ] I have reviewed CI checks for this PR and addressed them as far as I am able.

<!-- Thanks again for this PR - @epinowcast dev team -->

kaitejohnson commented:
Have broken this into a number of small functions but perhaps the logic itself is too complex?

seabbs commented on R/estimate_delay.R:
```diff
@@ -48,13 +48,29 @@ estimate_delay <- function(
     reporting_triangle,
     max_delay = ncol(reporting_triangle) - 1,
     n = nrow(reporting_triangle)) {
+  if (max_delay > ncol(reporting_triangle) - 1) {
```
This looks like it should error.

seabbs commented on R/estimate_delay.R:
```diff
@@ -48,13 +48,29 @@ estimate_delay <- function(
+    max_delay <- ncol(reporting_triangle) - 1
+  }
+
+  if (max_delay < ncol(reporting_triangle) - 1 && max_delay > 0) {
```
This will be quite an annoying message in practice. Is it needed or if it is is it needed for every single time it comes up i.e. do you want it once persession

kaitejohnson replied:
Probably once per session -- can change

kaitejohnson replied:
changed

seabbs commented on R/estimate_delay.R:
```diff
@@ -48,13 +48,29 @@ estimate_delay <- function(
+  if (max_delay < ncol(reporting_triangle) - 1 && max_delay > 0) {
+    # filter the reporting triangle to be less than the maximum delay
+    reporting_triangle <- reporting_triangle[, 1:(max_delay + 1)]
+    message(sprintf(
```
Why do you sometimes use cli tooling for messages, errors and warnings and sometimes not?

seabbs commented on tests/testthat/test-estimate_and_apply_uncertainty.R:
```diff
@@ -0,0 +1,97 @@
+    ),
+    regexp = "Using 6 reference times for delay estimation." # nolint
+  )
+  expect_identical(nowcast_draws_df, df_w_default)
```
Why are you testing it against itself like this?

kaitejohnson replied:
removed, replaced with tests against low level functions. Was just testing defaults and specification are same but weve moved to must specify

seabbs commented on R/allocate_reference_times.R:
```diff
@@ -0,0 +1,337 @@
+#' @inheritParams allocate_reference_times
+#'
+#' @returns NULL invisibly
+.validate_allocation_params <- function(scale_factor,
```
Fix. its likely the compound if else

kaitejohnson replied:
I simplified the compound if else since applies to all cases

seabbs commented on R/allocate_reference_times.R:
```diff
@@ -0,0 +1,337 @@
+    "Using %d reference times for delay estimation.",
+    n_history_delay
+  ))
+  message(sprintf("Using %d reference times as retrospective nowcast times for uncertainty estimation.", n_retrospective_nowcasts)) # nolint
```
what happened to cli?

kaitejohnson replied:
Will change to `cli_alert_info()`

seabbs commented on R/allocate_reference_times.R:
```diff
@@ -0,0 +1,337 @@
+#' @param prop_delay Numeric value <1 indicating what proportion of all
+#'     reference times in the reporting triangle to be used for delay
+#'     estimation. Default is `0.5`.
+#' @param size_min_retro_nowcasts Integer indicating the minimum number of
```
Why is this called size? What does it add as a prefix?

kaitejohnson replied:
Can remove the reference to size and just use `n`?

kaitejohnson replied:
The slight distinction in my head was `n` is user-specified, `size` is internally determined

seabbs commented on R/allocate_reference_times.R:
```diff
@@ -0,0 +1,337 @@
+#' Allocate training volume based on combination of defaults and user-specified
+#'   values for training volume for delay and uncertainty estimation.
+#' @description Given the reporting triangle, the maximum delay, and
```
How does it work? What is the alg? This needs to be documented.

seabbs commented on R/allocate_reference_times.R:
```diff
@@ -0,0 +1,337 @@
+  )
+
+  # If size being used equals or exceeds the target, simply split according to prop delay #nolint
+  ns <- .assign_ns_from_sizes(
```
i.e here if we didn't say size at all possible opporuntities we would have text that was 30% of the length

seabbs commented on R/allocate_reference_times.R:
```diff
@@ -0,0 +1,337 @@
+    n_retrospective_nowcasts <- size_used - n_history_delay
+    if (n_retrospective_nowcasts < size_min_retro_nowcasts) {
+      n_retrospective_nowcasts <- size_min_retro_nowcasts
+      n_history_delay <- size_used - n_retrospective_nowcasts
```
This line is duplicated

kaitejohnson replied:
I don't think the line is duplicated. This logic is implementing the case where your available reference times (size_used) are less than the target (size_target = max_delay*scale_factor)

If this is the case we first check that the delay estimation minimum is hit with:
` n_history_delay <- max(floor(size_used * prop_delay), size_min_delay)`
And if it is, then use prop delay to allocate. 

However, this needs a further check to ensure that when doing this, there are at least `size_min_retro_nowcasts`, if there aren't:
```r
 n_retrospective_nowcasts <- size_min_retro_nowcasts
 n_history_delay <- size_used - n_retrospective_nowcasts
```

kaitejohnson replied:
So we find delay first, assign the rest to uncertainty, and then check that uncertainty requirements are met.

seabbs commented on R/estimate_and_apply_delay.R:
```diff
@@ -0,0 +1,68 @@
+#' Estimate and apply delay from a reporting triangle
+#'
+#' @inheritParams estimate_delay
+#' @param ... Additional arguments
```
For what?

kaitejohnson replied:
This was requested in previous round of revisions so that we pass the default `n` from `estimate_delay()`.

seabbs replied:
no I mean how does the user know what this does as it isn't documented at all where the ... goes?

kaitejohnson replied:
Ah I see, have fixed to document

seabbs commented on R/validate.R:
```diff
@@ -168,6 +168,92 @@
+#' @inheritParams allocate_reference_times
+#' @autoglobal
+#' @returns NULL, invisibly
+.validate_ref_time_allocation <- function(n_ref_times,
```
This looks like a duplication of other checks but I am quite lost in the different validation parts of the code.

kaitejohnson replied:
There are two validation functions that have similar names. 

This one checks that the `n_history_delay` and `n_retrospective_nowcasts` passed into the `estimate_and_apply_delay()` function are valid -- I have renamed to `validate_inputs_uncertainty()`

The other checks that the parameters to `allocate_reference_times are valid, I have nammed this to `.validate_inputs_allocation`

seabbs replied:
and they share no code or overlap?

seabbs reviewed (changes requested):
Getting there. Some small issues - mostly code quality related.

kaitejohnson commented on R/estimate_and_apply_uncertainty.R:
```diff
@@ -0,0 +1,123 @@
+  )
+
+  # Logic to decide on how to allocate the training volume
+  n_ref_times <- nrow(reporting_triangle)
```
This is needed to validate the reference time allocation specified in the function `estimate_and_apply_delay()` ?

kaitejohnson replied:
I have removed the internal allocation and all defaults. Just a check that the specifications are valid.

kaitejohnson commented on R/validate.R:
```diff
@@ -168,6 +168,92 @@
+      "i" = "`ncol(reporting_triangle)` = {ncol(reporting_triangle)} but `max_delay + 1` = {max_delay + 1}." # nolint
+    ))
+  }
+
```
Have removed, will check files elsewhere...

seabbs commented on R/estimate_and_apply_delay.R:
```diff
@@ -0,0 +1,76 @@
+estimate_and_apply_delay <- function(reporting_triangle,
+                                     max_delay = ncol(reporting_triangle) - 1,
+                                     n_history_delay = NULL) {
+  if (max_delay > ncol(reporting_triangle) - 1) {
```
> Yeah but here I reset the maximum delay to be compatible with the reporting triangle. Do we not want to do this?

If we have the same check we should be sharing a validation function. No I don't think we want to do that.

kaitejohnson commented on R/estimate_and_apply_uncertainty.R:
```diff
@@ -0,0 +1,123 @@
+    n_ref_times = n_ref_times,
+    n_history_delay = n_history_delay,
+    n_retrospective_nowcasts = n_retrospective_nowcasts
+  ) # nolint
```
removed

kaitejohnson commented on R/estimate_and_apply_uncertainty.R:
```diff
@@ -0,0 +1,123 @@
+    truncated_reporting_triangles = trunc_rep_tris,
+    retro_reporting_triangles = retro_rep_tris,
+    n = n_retrospective_nowcasts,
+    uncertainty_model = uncertainty_model,
```
Unfortunately I don't think so because we are both estimating uncertainty and sampling nowcasts within this function, and this requires a different combination of these arguments... again this is related to eventually creating a constructor object which we can pass to both and then set as a default, but I think that should be in a separate PR see #229

kaitejohnson commented on tests/testthat/test-estimate_and_apply_delay.R:
```diff
@@ -0,0 +1,123 @@
+    reporting_triangle = reporting_triangle
+  )
+  # Test that the function returns the expected PMF
+  expect_equal(complete_triangle, point_nowcast_matrix, tol = 0.2)
```
will add

kaitejohnson commented on NAMESPACE:
```diff
@@ -20,6 +23,8 @@ export(truncate_triangles)
 importFrom(checkmate,assert_class)
 importFrom(checkmate,assert_integerish)
 importFrom(checkmate,assert_matrix)
+importFrom(checkmate,assert_numeric)
+importFrom(checkmate,assert_scalar)
```
checkmate is in imports

kaitejohnson commented:
Ok changes made:
- removed internal allocation in `estimate_and_apply_uncertainty()`
- streamlined logic within the assignment of `n_history_delay` and `n_retrospective_nowcasts` 
- added more comprehensive tests for the compound function (comparing to output from low level functions) and for allocation of reference times (testing scale factor and prop_delay work as expected)
- remove messages to cli_alert_info
- streamline max delay handling within both estimate_delay and estimate_and_apply_delay

kaitejohnson commented:
> Looking at this we should maybe change code rabbit to be called by the user as having in on every commit leaves a huge wall of code rabbit reviews to wade through. What do you think?

Yes -- this is why I lean heavily on the draft PR -- so it doesn't give me tons of comments while I'm iterating. Still an issue for multiple rounds of review though. I'd be fine with just calling upon it prior to asking for human review/when I think its reached a stable place.

seabbs commented:
I mean you shouldn't really be iterating on a non-draft PR anyway?
